core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala

```diff
@@ -0,0 +1,307 @@
+      assertEquals(s"Number of connections on $externalListener:", maxConnectionsPerIp + 2, connectionQuotas.get(externalListener.defaultClientIp))
+
+      // connections on the same listener but from a different IP should be accepted
+      executor.submit((() => acceptConnections(connectionQuotas, externalListener.listenerName, knownHost, maxConnectionsPerIp, 0)): Runnable).get(5, TimeUnit.SECONDS)
+
+      // remove two "rejected" connections and remove 2 more connections to free up the space for another 2 connections
+      for (_ <- 0 until 4) connectionQuotas.dec(externalListener.listenerName, externalListener.defaultClientIp)
+      assertEquals(s"Number of connections on $externalListener:", maxConnectionsPerIp - 2, connectionQuotas.get(externalListener.defaultClientIp))
+
+      executor.submit((() => acceptConnections(connectionQuotas, externalListener, 2)): Runnable).get(5, TimeUnit.SECONDS)
+      assertEquals(s"Number of connections on $externalListener:", maxConnectionsPerIp, connectionQuotas.get(externalListener.defaultClientIp))
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  def testMaxBrokerWideConnectionLimit(): Unit = {
+    val maxConnections = 800
+    val props = brokerPropsWithDefaultConnectionLimits
+    props.put(KafkaConfig.MaxConnectionsProp, maxConnections.toString)
+    val config = KafkaConfig.fromProps(props)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+
+    addListenersAndVerify(config, connectionQuotas)
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
+    try {
+      // the metric should be 0, because we did not advance the time yet
+      assertEquals(0, blockedPercentMeters("EXTERNAL").count())
+
+      // verify that ConnectionQuota can give all connections to one listener
+      // also add connections with a time interval between connections to make sure that time gets advanced
+      executor.submit((() => acceptConnections(connectionQuotas, listeners("EXTERNAL"), maxConnections, 1)): Runnable).get(5, TimeUnit.SECONDS)
+      assertEquals(s"Number of connections on ${listeners("EXTERNAL")}:", maxConnections, connectionQuotas.get(listeners("EXTERNAL").defaultClientIp))
+      // the blocked percent should still be 0, because there should be no wait for a connection slot
+      assertEquals(0, blockedPercentMeters("EXTERNAL").count())
+
+      // the number of connections should be above max for maxConnectionsExceeded to return true
+      assertFalse("Total number of connections is exactly the maximum.", connectionQuotas.maxConnectionsExceeded(listeners("EXTERNAL").listenerName))
+
+      // adding one more connection will block ConnectionQuota.inc()
+      val future = executor.submit((() => acceptConnections(connectionQuotas, listeners("EXTERNAL"), 1)): Runnable)
+      intercept[TimeoutException](future.get(1, TimeUnit.SECONDS))
```

Comment:
* You may want to assert here.
* I would reduce the waiting time to 10ms (or 100ms) to avoid always waiting 1s.